README.md
Bambora Client
==============

![Python application](https://github.com/archeti-org/bambora/workflows/Python%20application/badge.svg)
[![codecov](https://codecov.io/gh/archeti-org/bambora/branch/master/graph/badge.svg)](https://codecov.io/gh/archeti-org/bambora)
![PyPi](https://img.shields.io/badge/python-2.7%203.5-green.svg)

Simple client api that handle working with the REST
api from Bambora.

Example
=======


Create a new bambora object:

    bambora = Bambora({
        "merchant_id": "[Merchant ID]",
        "url": "https://api.na.bambora.com",
        "codes": {
            "payment_passcode":  "payment passcode",
            "payment_profile_passcode": "payment profile passcode",
            "report_passcode": "report passcode",
        }
    })

Create a token with:


    token = bambora.tokens.create({
        'number': "4030000010001234",
        'expiry_month': "10",
        'expiry_year': "22",
        'cvd': "123"
    })

Create a new payment profile with:

    profile = bambora.profiles.create({
        'token': {
            'name': 'Test',
            'code': token.token
        },
        'language': 'fr'
    })

List all the cards for this profile:

    cards = profile.cards.list()


Create a payment with the currently created profile:

    payment = bambora.payments.create({
        "order_number": "Some order id",
        "amount": 20,
        "payment_method": "payment_profile",
        "payment_profile": {
            "customer_code": profile.customer_code,
            "card_id": cards[0].get_id()
        }
    })


Delete objects:

    trans1 = cards[0].delete()
    trans2 = profile.delete()


CHANGELOG.md
2021-01-29:

- Check if the API still work as intended


bambora/tests/fixtures.py
# -*- coding: utf-8 -*-
import pytest
import os


@pytest.fixture
def bambora_env():
    from bambora import Bambora
    payment_profile = os.environ.get('PAYMENT_PROFILE_PASSCODE') 
    payment_report = os.environ.get('PAYMENT_REPORT_PASSCODE') 
    payment_gateway = os.environ.get('PAYMENT_GATEWAY_PASSCODE') 
    merchant_id = os.environ.get('BAMBORA_MERCHANT_ID')

    return Bambora({
        "merchant_id": merchant_id,
        "url": "https://api.na.bambora.com",
        "codes": {
            "payment_passcode": payment_gateway,
            "payment_profile_passcode": payment_profile,
            "report_passcode": payment_report,
        }
    })


bambora/tests/test_token.py
# -*- coding: utf-8 -*-
import pytest

import os
from bambora import Bambora
from bambora.objects import Token
from datetime  import datetime, timedelta
from bambora import exceptions

from .fixtures import bambora_env

def get_test_card_valid():
    year = (datetime.now() + timedelta(days=365*2)).strftime('%y')
    card = {
        'number': "4030000010001234",
        'expiry_month': "10",
        'expiry_year': year,
        'cvd': "123"
    }

    return card

def get_test_card_invalid():
    year = (datetime.now() + timedelta(days=-365*2)).strftime('%y')
    card = {
        'number': "4030000010001234",
        'expiry_month': "10",
        'expiry_year': year,
        'cvd': "321"
    }
    return card


def test_env_variables_present():
    assert os.environ.get('BAMBORA_MERCHANT_ID') is not None
    assert os.environ.get('PAYMENT_GATEWAY_PASSCODE') is not None
    assert os.environ.get('PAYMENT_PROFILE_PASSCODE') is not None
    assert os.environ.get('PAYMENT_REPORT_PASSCODE') is not None

def test_token_creation(bambora_env):
    token = bambora_env.tokens.create(get_test_card_valid())

    assert isinstance(token, Token)
    assert isinstance(token._attrs.get('token'), str)
    assert isinstance(token._attrs.get('token'), str)
    assert isinstance(token._attrs.get('version'), int)
    assert isinstance(token._attrs.get('message'), str)

    assert len(token._attrs.keys()) == 4


def test_create_profile(bambora_env):
    token = bambora_env.tokens.create(get_test_card_valid())
    profile = bambora_env.profiles.create({
        'token': {
            'name': 'Test',
            'code': token.token
        },
        'language': 'fr'
    })

    assert profile is not None

    cards = profile.cards.list()

    assert len(cards) == 1

    payment = bambora_env.payments.create({
        "order_number": "Test Trans (%s)" % (datetime.now().strftime('%Y%m%d%H%M%S%f'),),
        "amount": 20,
        "payment_method": "payment_profile",
        "payment_profile": {
            "customer_code": profile.customer_code,
            "card_id": cards[0].get_id()
        }
    })

    assert payment is not None

    trans1 = cards[0].delete()

    cards2 = profile.cards.list()

    assert len(cards2) == 0

    trans2 = profile.delete()

    with pytest.raises(exceptions.NotFoundError):
        trans3 = profile.delete()


examples/client.py
# -*- coding: utf-8 -*-
from bambora import Bambora
from bambora.exceptions import BamboraError, APIError
import logging

from argparse import ArgumentParser

logging.basicConfig(level=logging.INFO)

log = logging.getLogger(__name__)


parser = ArgumentParser()
parser.add_argument('--merchant', help="Merchant ID")
parser.add_argument('--payment', help="Payment Passcode")
parser.add_argument('--profile', help="Profile Passcode")
parser.add_argument('--reports', help="Reports Passcode")

args = parser.parse_args()

bambora = Bambora({
    "merchant_id": args.merchant,
    "url": "https://api.na.bambora.com",
    "codes": {
        "payment_passcode":  args.payment,
        "payment_profile_passcode": args.profile,
        "report_passcode": args.reports
    }
})

token = bambora.tokens.create({
    'number': "4030000010001234",
    'expiry_month': "10",
    'expiry_year': "22",
    'cvd': "123"
})

profile = bambora.profiles.create({
    'token': {
        'name': 'Test',
        'code': token.token
    },
    'language': 'fr'
})

cards = profile.cards.list()

payment = bambora.payments.create({
    "order_number": "ABC%s" % profile.customer_code[-5:],
    "amount": 20,
    "payment_method": "payment_profile",
    "payment_profile": {
        "customer_code": profile.customer_code,
        "card_id": cards[0].get_id()
    }
})

trans1 = cards[0].delete()
trans2 = profile.delete()


reports = []
try:
    fetch_data = True
    start = 1
    end = 10

    #while fetch_data:
    #    res = bambora.reports.list({
    #        'name': 'Search',
    #        'start_date': '2018-04-22T10:03:19',
    #        'end_date': '2018-10-03T10:03:19',
    #        'start_row': start,
    #        'end_row': end
    #    })
    #    if not len(res):
    #        fetch_data = False
    #    else:
    #        start += 20
    #        end += 20
    #        reports += res

    token2 = bambora.tokens.create({
        'number': "4030000010001235",
        'expiry_month': "10",
        'expiry_year': "22",
        'cvd': "123"
    })
    profile2 = bambora.profiles.create({
        "token": {
            "name": "Some name",
            "code": token2.token
        },
        "validate": True,
    })
except BamboraError as exc:
    import pdb; pdb.set_trace()
    log.info(exc.message)

#import pdb; pdb.set_trace()
#print(bambora.profiles.get('test').get_path())
#print(bambora.profiles.get_path())
